include/ibamr/FESurfaceDistanceEvaluator.h

```diff
@@ -0,0 +1,249 @@
+#include <vector>
+
+#include "PatchHierarchy.h"
+#include "ibamr/IBFEMethod.h"
+#include "ibtk/IndexUtilities.h"
+#include "ibtk/ibtk_utilities.h"
+#include "ibtk/libmesh_utilities.h"
+#include "libmesh/boundary_mesh.h"
+#include "tbox/Pointer.h"
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief class FESurfaceDistanceEvaluator is a utility class which is used to identify
+ * which line (triangle) in 2D (3D) elements are intersecting which grid cells.
+ */
+class FESurfaceDistanceEvaluator
+{
+public:
+    static const double s_large_distance;
+
+    /*!
+     * The only constructor of this class.
+     */
+    FESurfaceDistanceEvaluator(std::string object_name,
+                               SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                               SAMRAI::tbox::Pointer<IBAMR::IBFEMethod> ibfe_method,
+                               const libMesh::Mesh& mesh,
+                               const libMesh::BoundaryMesh& bdry_mesh,
+                               const int part,
+                               const int gcw = 1,
+                               bool use_extracted_bdry_mesh = true);
+
+    /*!
+     * Destructor for this class.
+     */
+    ~FESurfaceDistanceEvaluator();
+
+    /*!
+     * Map the triangles intersecting a particular grid cell.
+     */
+    void mapIntersections();
+
+    /*!
+     * Get the map maintaining triangle-cell intersection and neighbors.
+     */
+    const std::map<SAMRAI::pdat::CellIndex<NDIM>, std::set<libMesh::Elem*>, IBTK::CellIndexFortranOrder>&
+    getNeighborIntersectionsMap();
+
+    /*!
+     * Compute the signed distance in the viscinity of the finite element mesh.
+     */
+    void computeSignedDistance(int n_idx, int d_idx);
+
+    /*!
+     * Update the sign of the distance function away from the finite element mesh.
+     */
+    static void updateSignAwayFromInterface(int d_idx,
+                                            SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                                            double large_distance = s_large_distance);
+
+    /*!
+     * Check whether the grown box and line element intersect in 2D.
+     *
+     *     (box_tl)   *--------------*  (box_tr)
+     *                |              |
+     *                |              |
+     *                |              |
+     *                |              |
+     *                |              |
+     *     (box_bl)   *--------------*  (box_br)
+     */
+    static bool checkIntersection2D(const IBTK::Vector3d& box_bl,
+                                    const IBTK::Vector3d& box_tr,
+                                    const IBTK::Vector3d& box_br,
+                                    const IBTK::Vector3d& box_tl,
+                                    const libMesh::Point& n0,
+                                    const libMesh::Point& n1);
+
```

Comment:
Perhaps. I generally don't touch existing IBAMR code files, especially IBFE files to make my life easy.

Reply:
IBFE is too complicated for me to understand ☹️

Reply:
My point is not that we need to actually change IBFE now, but it will be easier to do so in the future if these functions are in the utility file instead of in this class. I just want to move theem.